[Libomptarget] Rework Record & Replay to be a plugin member (#88928)

jhuber6 opened the pull request:
Summary:
Previously, the R&R support was global state initialized by a global
constructor. This is bad because it prevents us from adequately
constraining the lifetime of the library. Additionally, we want to
minimize the amount of global state floating around.

This patch moves the R&R support into a plugin member like everything
else. This means there will be multiple copies of the R&R implementation
floating around, but this was already the case given the fact that we
currently handle everything with dynamic libraries.

jhuber6 commented:
New PR since GH is dumb and doesn't let you re-open a merged PR.

I tested this locally and couldn't reproduce any issues, I have no clue what could be going wrong. Could one of you test this locally and tell me where it's failing if it does?

jplehr commented:
I ran the tests locally and the segmentation fault reproduces. I randomly picked the `complex_reduction.cpp` test case to run in isolation and it appears to be a segmentation fault in the `deinit`. For reference: This machine has no CUDA installed, in case that makes a difference.

```
Thread 1 "complex_reducti" received signal SIGSEGV, Segmentation fault.
0x00007ffff103ebf9 in llvm::omp::target::plugin::GenericPluginTy::deinit() () from /path/bot-tester-builds/openmp/./lib/libomptarget.rtl.cuda.so
(gdb) bt
#0  0x00007ffff103ebf9 in llvm::omp::target::plugin::GenericPluginTy::deinit() () from /path/bot-tester-builds/openmp/./lib/libomptarget.rtl.cuda.so
#1  0x00007ffff1047982 in llvm::omp::target::plugin::PluginTy::deinit() () from /path/bot-tester-builds/openmp/./lib/libomptarget.rtl.cuda.so
#2  0x00007ffff104780d in llvm::omp::target::plugin::PluginTy::~PluginTy() () from /path/bot-tester-builds/openmp/./lib/libomptarget.rtl.cuda.so
```

jhuber6 commented:
As far as I can tell, the actual handling of this class is all over the place. It was a global reference that takes a single device and, is deinitialized along with the device, but must be initialized externally for a single device? I don't know, I just tried to retain the old semantics. Hopefully it works again.

jplehr commented:
I still get the segmentation fault error when I do a local build of this PR with our buildbot config.

jhuber6 committed:
[Libomptarget] Rework Record & Replay to be a plugin member

Summary:
Previously, the R&R support was global state initialized by a global
constructor. This is bad because it prevents us from adequately
constraining the lifetime of the library. Additionally, we want to
minimize the amount of global state floating around.

This patch moves the R&R support into a plugin member like everything
else. This means there will be multiple copies of the R&R implementation
floating around, but this was already the case given the fact that we
currently handle everything with dynamic libraries.

jplehr reviewed (commented):
LGTM
